shared-libs/rules-engine/src/task-fetcher.js

```diff
@@ -0,0 +1,210 @@
+  }
+
+  // Once the contact state store has all contacts, trust it and only refresh those marked dirty
+  return refreshTasksForKnownContacts(db, calculationTimestamp, contactStateStore.getContactIds());
+};
+
+const docsOf = query => query.then(result => result.rows.map(row => row.doc).filter(existing => existing));
+const fetch = {
+  allTaskData: db => {
+    const userDoc = contactStateStore.currentUser();
+    const userContactId = userDoc && userDoc._id;
+    return Promise.all([
+        docsOf(db.query('medic-client/contacts_by_type', { include_docs: true })),
+        docsOf(db.query('medic-client/reports_by_subject', { include_docs: true })),
+        fetch.allTasks(db, 'requester'),
+      ])
+      .then(([contactDocs, reportDocs, taskDocs]) => ({ contactDocs, reportDocs, taskDocs, userContactId }));
+  },
+
+  /*
+  PouchDB.query slows down when provided with a large keys array.
+  For users with ~1000 users it is ~50x faster to provider a start/end key instead of specifying all ids
+  */
+  allTasks: (db, prefix) => docsOf(db.query('medic-client/tasks', { startkey: `${prefix}-`, endkey: `${prefix}-\ufff0`, include_docs: true })),
+
+  existingContactStateStore: db => db.get(CONTACT_STATE_DOCID).catch(() => undefined),
+
+  tasksByRelation: (db, contactIds, prefix) => {
+    const keys = contactIds.map(contactId => `${prefix}-${contactId}`);
+    return docsOf(db.query('medic-client/tasks', { keys, include_docs: true }));
+  },
+
+  taskDataFor: (db, contactIds) => {
+    if (!contactIds || contactIds.length === 0) {
+      return Promise.resolve({});
+    }
+
+    const userDoc = contactStateStore.currentUser();
+    return docsOf(db.allDocs({ keys: contactIds, include_docs: true }))
+      .then(contactDocs => {
+        const subjectIds = contactDocs.reduce((agg, contactDoc) => {
+          registrationUtils.getSubjectIds(contactDoc).forEach(subjectId => agg.add(subjectId));
+          return agg;
+        }, new Set(contactIds));
+
+        return Promise.all([
+            docsOf(db.query('medic-client/reports_by_subject', { keys: Array.from(subjectIds).map(key => [key]), include_docs: true })),
+            fetch.tasksByRelation(db, contactIds, 'requester'),
+          ])
+          .then(([reportDocs, taskDocs]) => ({
+            userContactId: userDoc && userDoc._id,
+            contactDocs,
+            reportDocs,
+            taskDocs,
+          }));
+      });
+  },
+};
+
+// previousCommit helps avoid conflict errors since these changes are handled asynchronously
+let previousCommit = Promise.resolve();
+const stateChangeCallback = (db, stateDoc, updatedState) => {
+  Object.assign(stateDoc, { contactStateStore: updatedState });
+
+  previousCommit = previousCommit
+    .then(() => db.put(stateDoc))
+    .then(updatedDoc => { stateDoc._rev = updatedDoc.rev; })
+    .catch(err => console.error(`Error updating contactStateStore: ${err}`))
+    .then(() => { previousCommit = Promise.resolve(); });
```

Comment:
I added it because I wasn't sure of the memory implications of having a very long promise change.  Although previousCommit does resolve, if we keep adding promises to the chain could this bloat?